### DISCUSSION TITLE
SQLALchemy 2.0rc1 multi tenant migration (PostgreSQL schemas)

### DISCUSSION THREAD

**Discussion by mgurg** [sqlalchemy/alembic#1150] [Usage Questions] [answered]
Hi,
I'm using SQLALchemy with Alembic in the multitenant FastApi app. I would like to migrate to SQLAlchemy 2.0 but I have strange behaviour in migrations, where the only change is SQLAlchemy version: `1.4 -> 2.0.0rc1` 

The code is the same in both cases:
  -  `1.4` : https://github.com/mgurg/fastapi_docker/blob/sqlalchemy-14/requirements.txt
  -  `2.0.0rc1` : https://github.com/mgurg/fastapi_docker/blob/sqlalchemy-20/requirements.txt

In 2.0.0rc1 everything goes fine, but at the end I have `ROLLBACK` instead `COMMIT`, see attached files. Any ideas? I don't see any warnings or errors during code execution. Name of test PG schema: `fake_tenant_company_for_test_00000000000000000000000000000000`  

![SQ2_SQ14](https://user-images.githubusercontent.com/34517378/211358999-1c251531-44f5-4e73-812d-ff3935dc7983.png)

[SQ2.txt](https://github.com/sqlalchemy/sqlalchemy/files/10375219/SQ2.txt)
[SQ14.txt](https://github.com/sqlalchemy/sqlalchemy/files/10375220/SQ14.txt)

**Comment by CaselIT** [comment, marked as answer]
Hi,

There is some information that's missing here:
- how is alembic called
- what's the alembic env.py?
- what's the version of alembic?

I think this is more an alembic discussion, I can move it once the provided information is given if the issue turns out to be in alembic

**Reply by mgurg** [reply]
Alembic version: 1.9.1

I have this [endpoint](https://github.com/mgurg/fastapi_docker/blob/3c68a0e6b42a8125aefe9dbf46699d89f615a502/app/api/cc.py#L77)

which invokes `alembic_upgrade_head(tenant_id)`:

```python
def alembic_upgrade_head(tenant_name: str, revision="head", url: str = None):
    logger.info("🔺 [Schema upgrade] " + tenant_name + " to version: " + revision)
    print("🔺[Schema upgrade] " + tenant_name + " to version: " + revision)
    # set the paths values

    if url is None:
        url = SQLALCHEMY_DATABASE_URL
    try:
        # create Alembic config and feed it with paths
        config = Config(str(settings.PROJECT_DIR / "alembic.ini"))
        config.set_main_option("script_location", str(settings.PROJECT_DIR / "migrations"))  # replace("%", "%%")
        config.set_main_option("sqlalchemy.url", url)
        config.cmd_opts = argparse.Namespace()  # arguments stub

        # If it is required to pass -x parameters to alembic
        x_arg = "".join(["tenant=", tenant_name])  # "dry_run=" + "True"
        if not hasattr(config.cmd_opts, "x"):
            if x_arg is not None:
                setattr(config.cmd_opts, "x", [])
                if isinstance(x_arg, list) or isinstance(x_arg, tuple):
                    for x in x_arg:
                        config.cmd_opts.x.append(x)
                else:
                    config.cmd_opts.x.append(x_arg)
            else:
                setattr(config.cmd_opts, "x", None)

        # prepare and run the command
        revision = revision
        sql = False
        tag = None
        # command.stamp(config, revision, sql=sql, tag=tag)

        # upgrade command
        command.upgrade(config, revision, sql=sql, tag=tag)
    except Exception as e:
        logger.error(e)
        # capture_exception(e)
        # print(traceback.format_exc())

    logger.info("✅ Schema upgraded for: " + tenant_name + " to version: " + revision)
    print("✅ Schema upgraded for: " + tenant_name + " to version: " + revision)
```

`config.cmd_opts` is equal:

```python
Namespace(x=['tenant=fake_tenant_company_for_test_00000000000000000000000000000000'])
```

url is also correct: `postgresql+psycopg2://postgres:postgresp@***.***.***.***:5432/postgres`

### env.py
[github code](https://github.com/mgurg/fastapi_docker/blob/sqlalchemy-20/migrations/env.py)

I'm a beginner python programmer, if anything else is needed I will be glad to provide it.

**Reply by CaselIT** [reply]
thanks for the update.
the code seems fine to me. @zzzeek ideas?

**Reply by zzzeek** [reply]
for the code at https://github.com/mgurg/fastapi_docker/blob/fdb275533badb167f74c42e332bc2801dcd768bf/migrations/env.py#L83

change it to:

```py
    with connectable.connect() as connection:
        connection.execute(text('set search_path to "%s"' % current_tenant))
        connection.dialect.default_schema_name = current_tenant

        # add this line
        connection.commit()

        context.configure(
            connection=connection,
            target_metadata=target_metadata,
            version_table_schema=current_tenant,
        )
     
        with context.begin_transaction():
            context.run_migrations()
```

confirm that works and we will update the recipe on the site

**Reply by CaselIT** [reply]
is that because context gets a connection that's already in transaction, so it will not create and manage a new one by itself?

**Reply by zzzeek** [reply]
the call to ``connection.execute()`` is auto-beginning a transaction.  the ``context.begin_transaction()`` thing checks if the connection is in a trans already and does nothing if so.  similar subject as the Session changes we just did.

**Reply by mgurg** [reply]
With `connection.commit()` command DB is correctly populated with new tables. I attached logs, it looks like  `CREATE TABLE` statements are invoked always twice. This is not a problem for me at this moment, but maybe You would like to investigate this further.

[SQ2_with_fix.txt](https://github.com/sqlalchemy/alembic/files/10380006/SQ2_with_fix.txt)

![SQ2_with_fix](https://user-images.githubusercontent.com/34517378/211485496-c675ae7a-26f3-4d58-8ff9-6cbcd06372f9.png)

**Reply by CaselIT** [reply]
that seems just the logging module that has two handlers configured. Can you verify if that the case?

**Reply by mgurg** [reply]
You are 100% right, duplication is caused by those two handlers. With a single one everything looks fine.

[SQ2_with_fix_correct_logging.txt](https://github.com/sqlalchemy/alembic/files/10382272/SQ2_with_fix_correct_logging.txt)

**Reply by CaselIT** [reply]
thanks for testing!

>confirm that works and we will update the recipe on the site

@zzzeek what's the recipe to update?

**Reply by zzzeek** [reply]
that would be this one:  https://alembic.sqlalchemy.org/en/latest/cookbook.html#rudimental-schema-level-multi-tenancy-for-postgresql-databases

two approaches are:

1. add a connection.commit() call, and a comment "SQLAlchemy 2.0 only"
2. run the schema name change inside of a "with connection.begin()" block.

at the moment number one seems a little more clear to me what's going on.

**Reply by CaselIT** [reply]
I honestly like 2 better, since it's clear that we are committing the search path change

**Reply by CaselIT** [reply]
Created with https://gerrit.sqlalchemy.org/c/sqlalchemy/alembic/+/4355

I went with 1 since it's easier to mention that it's need on v2 only

**Comment by CaselIT** [comment]
moved to alembic
